// @vite-pwa/astro/dist/index.cjs
'use strict';Object.defineProperty(exports, '__esModule', {value: true});

const vitePluginPwa = require('vite-plugin-pwa');

function index(options = {}) {
  let pwaPlugin;
  const ctx = {
    preview: false,
    doBuild: false,
    scope: "/",
    trailingSlash: "ignore",
    useDirectoryFormat: true
  };
  const enableManifestTransform = () => {
    return ctx;
  };
  return {
    name: "@vite-pwa/astro-integration",
    hooks: {
      "astro:config:setup": ({ command, config, updateConfig }) => {
        if (command === "preview") {
          ctx.preview = true;
          return;
        }
        ctx.scope = config.base ?? config.vite.base ?? "/";
        ctx.trailingSlash = config.trailingSlash;
        ctx.useDirectoryFormat = config.build.format === "directory";
        let plugins = getViteConfiguration(
          config,
          options,
          ctx.useDirectoryFormat,
          enableManifestTransform
        );
        if (command === "build")
          plugins = plugins.filter((p) => "name" in p && p.name !== "vite-plugin-pwa:dev-sw");
        else
          plugins = plugins.filter((p) => "name" in p && p.name !== "vite-plugin-pwa:build");
        updateConfig({ vite: { plugins } });
      },
      "astro:config:done": ({ config }) => {
        if (ctx.preview)
          return;
        pwaPlugin = config.vite.plugins.flat(Number.POSITIVE_INFINITY).find((p) => p.name === "vite-plugin-pwa");
      },
      "astro:build:done": async () => {
        if (ctx.preview)
          return;
        ctx.doBuild = true;
        await regeneratePWA(pwaPlugin);
      }
    }
  };
}
function createManifestTransform(enableManifestTransform) {
  return async (entries) => {
    const { doBuild, trailingSlash, scope, useDirectoryFormat } = enableManifestTransform();
    if (!doBuild)
      return { manifest: entries, warnings: [] };
    entries.filter((e) => e && e.url.endsWith(".html")).forEach((e) => {
      const url = e.url.startsWith("/") ? e.url.slice(1) : e.url;
      if (url === "index.html") {
        e.url = scope;
      } else {
        const parts = url.split("/");
        parts[parts.length - 1] = parts[parts.length - 1].replace(/\.html$/, "");
        e.url = useDirectoryFormat ? parts.length > 1 ? parts.slice(0, parts.length - 1).join("/") : parts[0] : parts.join("/");
        if (trailingSlash === "always")
          e.url += "/";
      }
    });
    return { manifest: entries, warnings: [] };
  };
}
function createExperimentalManifestTransform(enableManifestTransform) {
  return async (entries) => {
    const { doBuild, trailingSlash, scope, useDirectoryFormat } = enableManifestTransform();
    if (!doBuild)
      return { manifest: entries, warnings: [] };
    const additionalEntries = [];
    entries.filter((e) => e && e.url.endsWith(".html")).forEach((e) => {
      const url = e.url.startsWith("/") ? e.url.slice(1) : e.url;
      if (url === "index.html") {
        additionalEntries.push({
          revision: e.revision,
          url: scope,
          size: e.size
        });
      } else if (url === "404.html") {
        e.url = `404${trailingSlash === "always" ? "/" : ""}`;
      } else {
        const parts = url.split("/");
        parts[parts.length - 1] = parts[parts.length - 1].replace(/\.html$/, "");
        let newUrl = useDirectoryFormat ? parts.length > 1 ? parts.slice(0, parts.length - 1).join("/") : parts[0] : parts.join("/");
        if (trailingSlash === "always")
          newUrl += "/";
        additionalEntries.push({
          revision: e.revision,
          url: newUrl,
          size: e.size
        });
      }
    });
    if (additionalEntries.length)
      entries.push(...additionalEntries);
    return { manifest: entries, warnings: [] };
  };
}
function getViteConfiguration(config, options, directoryFormat, enableManifestTransform) {
  const plugin = config.vite?.plugins?.flat(Number.POSITIVE_INFINITY).find((p) => p.name === "vite-plugin-pwa");
  if (plugin)
    throw new Error("Remove the vite-plugin-pwa plugin from Vite Plugins entry in Astro config file, configure it via @vite-pwa/astro integration");
  options.includeManifestIcons = false;
  const {
    strategies = "generateSW",
    registerType = "prompt",
    injectRegister,
    workbox = {},
    ...rest
  } = options;
  let assets = config.build.assets ?? "_astro/";
  if (assets[0] === "/")
    assets = assets.slice(1);
  if (assets[assets.length - 1] !== "/")
    assets += "/";
  if (strategies === "generateSW") {
    const useWorkbox = { ...workbox };
    const newOptions = {
      ...rest,
      strategies,
      registerType,
      injectRegister
    };
    if (!("navigateFallback" in useWorkbox))
      useWorkbox.navigateFallback = config.base ?? config.vite?.base ?? "/";
    if (directoryFormat)
      useWorkbox.directoryIndex = "index.html";
    newOptions.workbox = useWorkbox;
    if (!("dontCacheBustURLsMatching" in newOptions.workbox))
      newOptions.workbox.dontCacheBustURLsMatching = new RegExp(assets);
    if (!newOptions.workbox.manifestTransforms) {
      newOptions.workbox.manifestTransforms = newOptions.workbox.manifestTransforms ?? [];
      newOptions.workbox.manifestTransforms.push(
        options.experimental?.directoryAndTrailingSlashHandler === true ? createExperimentalManifestTransform(enableManifestTransform) : createManifestTransform(enableManifestTransform)
      );
    }
    return vitePluginPwa.VitePWA(newOptions);
  }
  options.injectManifest = options.injectManifest ?? {};
  if (!("dontCacheBustURLsMatching" in options.injectManifest))
    options.injectManifest.dontCacheBustURLsMatching = new RegExp(assets);
  if (!options.injectManifest.manifestTransforms) {
    options.injectManifest.manifestTransforms = options.injectManifest.manifestTransforms ?? [];
    options.injectManifest.manifestTransforms.push(
      options.experimental?.directoryAndTrailingSlashHandler === true ? createExperimentalManifestTransform(enableManifestTransform) : createManifestTransform(enableManifestTransform)
    );
  }
  return vitePluginPwa.VitePWA(options);
}
async function regeneratePWA(pwaPlugin) {
  const api = pwaPlugin?.api;
  if (api && !api.disabled) {
    await api.generateSW();
  }
}

exports.default = index;
